### DISCUSSION TITLE
Compose database URL from several ENV variables in schema.prisma file

### DISCUSSION THREAD

**Discussion by MurzNN** [prisma/prisma#12600] [Q&A] [answered]
I have database access credentials in separate ENV variables like `DB_USERNAME`, `DB_PASSSWORD`, `DB_DATABASE`, but in `schema.prisma` file Prisma wants to see a full url like this:
```
datasource db {
  provider = "postgresql"
  url      = "postgresql://johndoe:mypassword@localhost:5432/my_database"
}
```
So, can I use environment variables as part of `url` string, like this `"postgresql://$DB_USERNAME:$DB_PASSWORD@localhost:5432/$DB_DATABASE"`?

If not, is there any other ways to compose database url from several ENV variables?

**Comment by MurzNN** [comment]
It is possible to use a single ENV variable as database url like this:
```
datasource postgresql {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}
```
But does it work for concatenate several env variables?

**Comment by nurul3101** [comment, marked as answer]
Hey @MurzNN 👋 
Are you looking to do something like this - [Expanding Variables in env file](https://www.prisma.io/docs/guides/development-environment/environment-variables#expanding-variables)?

Another alternative is to override database URL programmatically as mentioned [here](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#programmatically-override-a-datasource-url). 
You could try constructing your database url before passing it in the `PrismaClient` Instance.

**Reply by MurzNN** [reply]
I'm already overriding database URL programmatically and it works for my application well, but the problem is that `prisma migrate deploy` command looks only to the `prisma.schema` file and fails on connecting step.

Yes, workaround is to duplicate database credentials and compose `$DATABASE_URL` string manually, but I'm not using `.env` file, all environment variables are passed to the container from Kubernetes, so better is to resolve this without adding new env variable to the infrastructure.

**Reply by janpio** [reply]
Have a look at the first link @nurul3101 posted @MurzNN - that is a workaround to compose variables via an `.env` file. That shuld also work for CLI commands (opposed to doing it in code).

**Reply by MurzNN** [reply]
Yes, workaround also is to fill the env variable directly with launching prisma:
```
DATABASE_URL="postgresql://$DB_USERNAME:$DB_PASSWORD@localhost:5432/$DB_DATABASE" prisma migrate deploy
```

**Reply by rolandboon** [reply]
The relevant docs mentioned by @nurul3101 has been moved to the following URL: https://www.prisma.io/docs/orm/more/development-environment/environment-variables/env-files#expanding-variables

**Reply by sovetski** [reply]
> Yes, workaround also is to fill the env variable directly with launching prisma:
> 
> ```
> DATABASE_URL="postgresql://$DB_USERNAME:$DB_PASSWORD@localhost:5432/$DB_DATABASE" prisma migrate deploy
> ```

it does not work for me.. the variables are not detected

**Comment by Telokis** [comment]
It's been almost three years, is there any way to solve this without manually specifying the URL?  
I'm not relying on `.env` files either, I would like `prisma migrate` to match the way my app is configured.

Ideally a way to programmatically perform `prisma migrate` operations would facilitate this process. In the meantime I think my solution will be to make a script that loads my config management and then instantiates `prisma migrate` as a child process with the URL set as an environment variable. Quite complicated.

By forcing us to use a database url instead of the individual components, you make your own work easier but end users will never manually type out a database url, it's quite complex for no gain. I let my users specify the database username, password, host, port and database name by themselves and I provide sensible defaults to make things as smooth as possible for them, instead.

Here is what it looks like in the app code:
```ts
import { PrismaClient } from "@prisma/client";
import config from "config";

export const prisma = new PrismaClient({
	datasourceUrl: [
		"postgresql://",
		config.postgres.username,
		":",
		config.postgres.password,
		"@",
		config.postgres.host,
		":",
		config.postgres.port,
		"/",
		config.postgres.database,
		"?schema=public",
	].join(""),
});

```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
